### PR TITLE
shelley update mechanism

### LINKED ISSUE #448: Scope work required to adapt the Byron Update Mechanim for Shelley.
In Byron, there is one update payload per block. In Shelley there is currently one per transaction. There is a problem, however, with using more than one `end ∈ VKey × ProtVer` per block. This component will probably have to be moved to the block body. This may mean having to redo how Shelley uses the Bryon rules.

In particular, in Shelley, we need to distinguish between 1) proposals and votes, and 2) endorsements. How this is handled will depend on how the mapping between genesis keys and cold keys is handled #447 Either the `UPEC` or `UPIEC` Byron rule will need to be handled separately for Shelley

Scope out the required work for adapting the update mechanism to Shelley.

### PR DESCRIPTION
**Pull request by JaredCorduan**
This PR introduces an update mechanism for the Shelley ledger rules.

There is a new section, `7. Updates` which describes the update mechanism.

Transaction now contain a special component available only to the delegates of the genesis keys. It consists of a pair, one part for protocol parameter updates and one for application versions. I've not made application versions a protocol parameter since they are processed quite differently.  In particular, protocol parameter updates happen only at the epoch boundary, but application versions change "2k" slots after consensus is reached (we do not yet know if the value is really 2k, it is called `SlotsPrior` in the spec).

The new section describes three new transition systems. The first `PPUP` handles the protocol parameters, the second `AVUP` handles the application versions, and `UP` combines them.

The `AVUP` is pretty self contained, but the state in `PPUP` is used in the `EPOCH` transition (fig 46) to get new protocol parameters.

There is also another new section `13 Software Updates` that briefly describes how software updates work.

Before we can really say that the "shelley update mechanism" is done, we probably need to add more prose. Moreover, we need more checks around new protocol parameters, for which I've just made #572.

closes #448

### CODE REVIEW

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 1]
```diff
@@ -625,24 +626,28 @@ \subsection{Protocol Parameter Update Transition}
 
 Figure~\ref{fig:rules:new-proto-param} defines the new protocol parameter transition.
 The transition has two rules, depending on whether or not the new protocol parameters
-would incur a debt of the system that could not be covered by the reserves.
-The transition has two rules, each with one precondition. The preconditions are
-negations of each other, so that exactly one will always be met.
-This transition results in the following state change:
+meet some requirements.
+In particular, we require that the new parametern would not incur a debt of the system that
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]:
```suggestion
In particular, we require that the new parameters would not incur a debt of the system that
```

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 2]
```diff
@@ -625,24 +626,28 @@ \subsection{Protocol Parameter Update Transition}
...
+can not be covered by the reserves, and that the max block size is great than the sum of the
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]:
```suggestion
can not be covered by the reserves, and that the max block size is greater than the sum of the
```

Let's make the max block size greater again :clown_face:

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 3]
```diff
@@ -625,24 +626,28 @@ \subsection{Protocol Parameter Update Transition}
...
+max transaction size and the max header size.
+If the requirements are met, the new protocol parameters are accepted, the proposal is reset,
+and the reserves are adjusted to account for changes in the deposits.
+Otherwise, the only change is that the proposal is reset.
+
+Regarding adjusting the reseverse for changes in the deposits, one of three things happens:
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]: Reserves?

File: shelley/chain-and-ledger/formal-spec/updates.tex [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+\label{sec:updates}
+
+This section describes how software updates are performed.
+The process of upgrading the system a new version consists of:
+
+\begin{enumerate}
+  \item New software is ready for downloading. The core nodes propose changing the 
+    \textit{software version} in the protocol parameters. If accepted, the proposal
+    will change the protocol parameters at the start of the next epoch.
+  \item All nodes see a new software version and (hopefully) update their software.
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/updates.tex]: I don't know if we should say "hopefully" here ...

File: shelley/chain-and-ledger/formal-spec/updates.tex [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    Note that the new software will be compatible with the older version, and no changes to
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/updates.tex]: Is there something that prevents the new software from being incompatible with the old version? If not maybe we should rephrase this as saying "it should be compatible with the old version" :man_shrugging:

File: shelley/chain-and-ledger/formal-spec/updates.tex [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    the ledgers rules will occur as long as the protocol version in the protocol parameters
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/updates.tex]: hmm... maybe "the new ledger rules will not be applied/used as long as ..."

File: shelley/chain-and-ledger/formal-spec/updates.tex [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    is not increased.
+  \item Every block producer includes the protocol version corresponding to the softare
+    they are using in the blocks they produce.
+  \item The core nodes monitor how many block headers contain the new protocol version.
+    Once enough blocks have the new version, the core nodes post another protocol parameter
+    update to increas the protocol version.
+    If this proposal is accepted, it will be changed at the start of the new epoch and
+    the current software then makes use of new rules.
+\end{enumerate}
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/updates.tex]: Do we need a similar section describing protocol parameter updates? And maybe the meaning of protocol versions?

**Comment by dnadales** [comment]
image

Are we still using these?

**Comment by dnadales** [comment]
image

Should `s` be `slot`? Or vice-versa ...

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 8]
```diff
@@ -726,9 +734,26 @@ \subsection{Protocol Parameter Update Transition}
           \var{oblg_{cur}} & \leteq & \obligation{pp}{stkeys}{stpools}{slot} \\
           \var{oblg_{new}} & \leteq & \obligation{pp_{new}}{stkeys}{stpools}{slot} \\
           \var{diff} & \leteq & \var{oblg_{cur}} - \var{oblg_{new}} \\
-          \\
-          \var{reserves} + \var{diff} & < & 0\\
-      \end{array}}\\
+      \end{array}}
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]: BTW, shouldn't this section be called "Protocol Parameters Update Transition"?

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 9]
```diff
@@ -810,6 +825,31 @@ \subsection{Complete Epoch Boundary Transition}
...
+  \begin{align*}
+      & \fun{pvotes} \in \Update \to \Ppm \to (\VKeyGen \mapsto \Value) \\
+      & \fun{pvotes}~\var{up}~\var{ppm}
+      = \{k\mapsto v~\mid~ k\mapsto\var{ppm} ~\land~ \var{ppm}\mapsto v\}\\
+  \end{align*}
+
+  \begin{align*}
+      & \fun{tally} \in \Update \to \PParams\\
+      & \fun{tally}~\var{up} =
+          \Big\{\var{ppm}\mapsto v
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]: It seems `v` is not bound to anything on the rhs of the "pipe":

image

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 10]
```diff
@@ -810,6 +825,31 @@ \subsection{Complete Epoch Boundary Transition}
...
+$\mathsf{EPOCH}$ transition.
+
+%%
+%% Figure - Helper Functions for Epoch Rules
+%%
+\begin{figure}[htb]
+  \begin{align*}
+      & \fun{pvotes} \in \Update \to \Ppm \to (\VKeyGen \mapsto \Value) \\
+      & \fun{pvotes}~\var{up}~\var{ppm}
+      = \{k\mapsto v~\mid~ k\mapsto\var{ppm} ~\land~ \var{ppm}\mapsto v\}\\
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]: Do you mean `k \mapsto ppm \in up`? (and the same for `v`?)

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 11]
```diff
@@ -810,6 +825,31 @@ \subsection{Complete Epoch Boundary Transition}
...
+  \end{align*}
+
+  \begin{align*}
+      & \fun{tally} \in \Update \to \PParams\\
+      & \fun{tally}~\var{up} =
+          \Big\{\var{ppm}\mapsto v
+          ~\Big|~
+          |\fun{pvotes}~\var{up}~\var{ppm}\restrictrange\var{v}|\geq 5 \Big\}
+  \end{align*}
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]: I'm also wondering, do we need the flexibility of updating individual values? Can't we just say that we have a mapping `VKeyG \mapsto Ppm` and then simply compare (and count) the `Ppm` maps?

File: shelley/chain-and-ledger/formal-spec/epoch.tex [hunk 12]
```diff
@@ -810,6 +825,31 @@ \subsection{Complete Epoch Boundary Transition}
...
+
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/epoch.tex]: I don't know if `tally` is well defined in the case that you have a protocol parameter with two majorities (which if we have 7 keys shouldn't happen, but still the definition of `tally` doesn't now this ;) )

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 3 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: yes!

File: shelley/chain-and-ledger/formal-spec/updates.tex (hunk 4 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/updates.tex]: yea, good point, I'll remove it.

File: shelley/chain-and-ledger/formal-spec/updates.tex (hunk 5 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/updates.tex]: hmm, ok, that's fair. kinda like my "hopefully" above :)

File: shelley/chain-and-ledger/formal-spec/updates.tex (hunk 7 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/updates.tex]: yea, I do think the current version is short on prose for the updates.

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 8 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: Yes, that 's probably consistent with the rest.

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 9 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: it _could_ just be any `v` that type checks, but in practice it would be the values of some protocol parameter in the update. I'll try to reword it better...

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 11 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: yea, that's fine too. both are equal easy to write and implement, one just gives you more options. I'm happy either way.

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 12 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: I'm making the assumption that there are only ever seven keys. maybe I can state that somewhere.

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 8 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: oh wait, I misunderstood, you mean "parameters" instead of "parameter", right?

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 9 above)
Comment (dnadales) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: Hmmmm, but then we'd need an existential in the range of the set comprehension ... I guess `v` can only come from `range (range up)`, right? But anyway, I'll wait for the rewording :D

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 11 above)
Comment (dnadales) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: Same here, but I think that this flexibility makes computation of agreement a bit more complex.

**Comment by dnadales** [comment]
BTW, you might want to check these:

```
Latexmk: List of undefined refs and citations:
  Label `fig:funcs:epoch-helper' multiply defined
  Label `fig:funcs:epoch-helper@cref' multiply defined
```

**Comment by dnadales** [comment]
One thing I forgot to ask, do we check that the update proposal is stable before counting them? Otherwise what happens in the case of a rollback?

image

File: shelley/chain-and-ledger/formal-spec/epoch.tex (hunk 8 above)
Comment (dnadales) [reply shelley/chain-and-ledger/formal-spec/epoch.tex]: Right, aren't we updating a bunch of parameters? But I don't know whether that's the right wording :man_shrugging: :es:

**Comment by JaredCorduan** [comment]
> One thing I forgot to ask, do we check that the update proposal is stable before counting them? Otherwise what happens in the case of a rollback?

good question @dnadales.  New updates cannot be supplied near the end of the epoch:

up-ne

`SlotsPrior` is the same constant used to ensure that the Praos nonce is stable at the beginning of the next epoch. Sometimes I call it "2k", but I think the precise value is being worked out.

Does this relieve your concern?

**Comment by dnadales** [comment]
> Does this relieve your concern?

Yes :)

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 13]
```diff
@@ -0,0 +1,396 @@
+\section{Updates}
+\label{sec:update}
+
+
+The $\mathsf{UPDATE}$ transition is responsible for federated governance model in Shelly.
+This includes the ability to change the protocol parameters and also handles software
+updates.
+
+The application versions behave much like a protocol parameter, but because adoption
+for them must occur as soon as consensus is met, they are handled separately.
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: maybe:

```latex
as soon as consensus is met\footnote{as opposed to waiting till the end of an epoch}
```

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+Therefore the $\mathsf{UPDATE}$ is the combination of the
+Therefore the $\mathsf{PPUP}$ transition for the protocol parameters and the
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: Therefore we have one `Therefore` too many :clown_face:

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        \var{dms} & \VKeyGen\mapsto\VKey & \text{genesis key delegations} \\
+      \end{array}
+    \right)
+  \end{equation*}
+  %
+  \emph{Protocol Parameter Update transitions}
+  \begin{equation*}
+    \_ \vdash
+    \var{\_} \trans{ppup}{\_} \var{\_}
+    \subseteq \powerset (\PPUpdateEnv \times \PPUpdate \times \PPUpdate \times \PPUpdate)
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: Would it make sense to include a hyperlinked reference here to where `\PPUpdate` is defined?

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+  \end{equation}
+
+  \nextdef
+
+  \begin{equation}\label{eq:update-nonempty}
+    \inference[PP-Update-Nonempty]
+    {
+      \var{pup}\neq\emptyset
+      &
+      \dom{pup}\subseteq\dom{dms}
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: It seems that we use `dms` here only for its domain. Would it make sense to pass in the environment only what's needed, i.e. the allowed proposers (?).

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+
+%%
+%% Figure - Helper Function for Consensus of Update Proposals
+%%
+\begin{figure}[htb]
+  \begin{align*}
+      & \fun{votedValue_T} \in (\VKeyGen\mapsto\type{T}) \to \type{T}^?\\
+      & \fun{votedValue_T}~\var{vs} =
+        \begin{cases}
+          t & \exists t\in\range{vs}(|vs\restrictrange t|\geq 5) \\
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: It seems we miss a `\cdot` between the range of the quantification and its term.

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+genesis keys agree.
+This function will also be used for the protocol parameters later in Section~\ref{sec:epoch}.
+Note that $\type{T}$ is an arbitrary type.
+
+%%
+%% Figure - Helper Function for Consensus of Update Proposals
+%%
+\begin{figure}[htb]
+  \begin{align*}
+      & \fun{votedValue_T} \in (\VKeyGen\mapsto\type{T}) \to \type{T}^?\\
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: Is `votedValue_T` well defined in the case there are multiple values that satisfy the predicate?

Why not defining the values for which there's an agreement as:

```
{ v | v ∈ range vs, 5 <= |vs ▷ v|  }
```

But of course the question is how do we plan on using this... I'll read on ...

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+      & \fun{votedValue_T}~\var{vs} =
+        \begin{cases}
+          t & \exists t\in\range{vs}(|vs\restrictrange t|\geq 5) \\
+          \Nothing & \text{otherwise} \\
+        \end{cases}
+  \end{align*}
+
+  \begin{align*}
+      & \fun{newAVs} \in \Applications \to (\Slot\mapsto\Applications) \to \Applications \\
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: Maybe we need to refer in the prose to the place where `Applications` is defined.

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+      &
+      \dom{\var{aup}}\subseteq\dom{\var{dms}}
+      \\
+      \var{aup'}\leteq\var{aup_s}\unionoverrideRight\var{aup}
+      &
+      \var{fav}\leteq\fun{votedValue_{Applications}}~\var{aup'}
+      \\
+      \var{fav}=\Nothing
+      \\
+      \var{cur}\leteq\{\var{s}\mapsto\var{v}\in\var{favs} ~\mid~ s\geq\var{slot}\}
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: We miss a let binding for `s`

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 21]
```diff
@@ -0,0 +1,396 @@
...
+  \end{equation*}
+  %
+  \emph{Application Version Update states}
+  \begin{equation*}
+    \AVUpdateState =
+    \left(
+      \begin{array}{r@{~\in~}lr}
+        \var{aup} & \AVUpdate & \text{application versions proposals} \\
+        \var{favs} & \Slot\mapsto\Applications & \text{future application versions} \\
+        \var{avs} & \Applications & \text{current application versions} \\
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: In the Byron spec @dcoutts pointed out to me that we needed to keep the slot in which the application version changed. We might need the same here.

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 22]
```diff
@@ -0,0 +1,396 @@
...
+      \begin{array}{l}
+        \var{aup_s}\\
+        \var{favs}\\
+        \var{avs}\\
+      \end{array}
+      \right)
+      \trans{avup}{aup}
+      \left(
+      \begin{array}{l}
+        \varUpdate{\emptyset}\\
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: Here it is kind of weird that we throw away the other proposals when there is consensus on a given application ...

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 23]
```diff
@@ -0,0 +1,396 @@
...
+  %
+  \caption{Application Version Update transition-system types}
+  \label{fig:ts-types:update}
+\end{figure}
+
+\begin{figure}[htb]
+  \begin{equation}\label{eq:update}
+    \inference[Update]
+    {
+      (\var{pup},~\var{aup})\leteq\var{up}
```
Comment (dnadales) [inline shelley/chain-and-ledger/formal-spec/update.tex]: Here it seems we're forcing an update to contain parameters and software updates. Do we need this constraint?

File: shelley/chain-and-ledger/formal-spec/update.tex (hunk 21 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/update.tex]: I think `versions` is correct, since there are multiple `AppName`s

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 24]
```diff
@@ -0,0 +1,396 @@
+\section{Updates}
+\label{sec:update}
+
+
+The $\mathsf{UPDATE}$ transition is responsible for federated governance model in Shelly.
+This includes the ability to change the protocol parameters and also handles software
+updates.
+
+The application versions behave much like a protocol parameter, but because adoption
```
[the remark this answers is not in the document]
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/update.tex]: yes. I was trying to use the phrase "software update" for the actual updating of software, and "application version" to refer to what the software will be updating too. but I may have been inconsistent...

File: shelley/chain-and-ledger/formal-spec/transactions.tex [hunk 25]
```diff
@@ -73,10 +72,25 @@ \section{Transactions}
...
-      & \VKeyGen \mapsto \Seed
-      & \text{extra entropy}
+      \var{pup}
+      & \PPUpdate
+      & \VKeyGen \mapsto \Ppm \mapsto \Seed
+      & \text{protocol parameter update}
+      \\
+      \var{apps}
+      & \Applications
+      & \ApName \mapsto (\ApVer \times \Metadata)
```
[the remark this answers is not in the document]
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/transactions.tex]: ah, not in this spec, thanks! I will link to the Byron spec.

File: shelley/chain-and-ledger/formal-spec/update.tex (hunk 16 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/update.tex]: I'm going to pass on this for now, if that's okay. once all the dust settles, we can think about if this optimization is worth it.

File: shelley/chain-and-ledger/formal-spec/update.tex (hunk 21 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/update.tex]: @dnadales I made this issue so that we don't forget: #575

File: shelley/chain-and-ledger/formal-spec/update.tex (hunk 24 above)
[the remark this answers is not in the document]
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/update.tex]: ok, I added some prose, thanks!

File: shelley/chain-and-ledger/formal-spec/transactions.tex (hunk 25 above)
[the remark this answers is not in the document]
Comment (JaredCorduan) [reply shelley/chain-and-ledger/formal-spec/transactions.tex]: nm, it's easier to just define them directly, since it's just two abstract types and one alias for N.

File: shelley/chain-and-ledger/formal-spec/update.tex (hunk 16 above)
Comment (dnadales) [reply shelley/chain-and-ledger/formal-spec/update.tex]: Sure.

File: shelley/chain-and-ledger/formal-spec/update.tex (hunk 21 above)
Comment (dnadales) [reply shelley/chain-and-ledger/formal-spec/update.tex]: Great! Thanks!
